io-engine/src/rebuild/rebuild_log.rs

```diff
@@ -0,0 +1,124 @@
+use bit_vec::BitVec;
+use parking_lot::Mutex;
+use spdk_rs::IoType;
+use std::{
+    fmt::{Debug, Formatter},
+    sync::Arc,
+};
+
+use super::SEGMENT_SIZE;
+
+// Returns ceil of an integer division.
+fn div_ceil(a: u64, b: u64) -> u64 {
+    (a + b - 1) / b
+}
+
+/// Rebuild log data structure.
+pub struct RebuildLog {
+    /// Name of the underlying device, stored for diagnostic purposes.
+    device_name: String,
+    /// Segment bitmap. Ones indicate modified segments.
+    segments: BitVec,
+    /// Device size in segments.
+    num_segments: u64,
+    /// Device size in block.
+    num_blocks: u64,
+    /// Size of block in bytes.
+    block_len: u64,
+}
+
+impl Debug for RebuildLog {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        write!(
+            f,
+            "Rebuild log: '{}' ({} segments: {} blocks x {})",
+            self.device_name,
+            self.num_segments,
+            self.num_blocks,
+            self.block_len
+        )
+    }
+}
+
+impl RebuildLog {
+    /// Creates a new rebuild log instance for the given device.
+    pub fn new(device_name: &str, num_blocks: u64, block_len: u64) -> Self {
+        let num_segments = div_ceil(num_blocks * block_len, SEGMENT_SIZE);
```

Comment:
Can we avoid this calculation everytime by simply using size_in_bytes from BlockDevice trait?

Reply:
There's no problem calculating it here, as it is executed only once when a log is created. Even on I/O path, a single division won't hurt: our code is not CPU bound but rather I/O bound.
From the design perspective, "segment" is rebuild's job thing, block devices know nothing about how rebuild algorithm does its copy job.

Reply:
We're already cloning a string anyway, but I'm not sure it will always be I/O bound as we're talking about nvme and you could have multiple pools.
But anyway this is only for the creation so it won't happen very often.

Reply:
Division on modern CPUs is nothing, unless you have to process 8K 60fps video in realtime :-)